tests/config/config.d/storage_conf.xml

```diff
@@ -6,6 +6,7 @@
                 <endpoint>http://localhost:11111/test/00170_test/</endpoint>
                 <access_key_id>clickhouse</access_key_id>
                 <secret_access_key>clickhouse</secret_access_key>
+                <data_cache_method>ARC</data_cache_method>
```

Comment:
> I would also rather copy tests where cache is tested or add randomization, so both methods are tested in CI. Lets copy tests to include both policies. To make it easier, let's add a setting in Settings.h which will define a default cache policy. It will be used if there is no value `cache_method` in disk config (instead of hardcoded `"ARC"` in `config.getString(config_prefix + ".data_cache_method", "ARC");`). So let's add it for the following tests:
> 
> ```
> stateless:
> 02241_remote_filesystem_cache_on_insert.sql
> 02226_s3_with_cache.sql
> 02240_system_remote_filesystem_cache.sql
> 
> stateful:
> 00170_s3_cache.sql
> ```
> 
> Probably it could be helpful to make an `.sh` instead of these `.sql` tests (see `0_stateless`, there are many `.sh` tests). So in this `.sh` test there can be a loop over a list of cache policies (for now only two). For each policy it will run clickHouse-client command with `--queries-file`. In the file let's put the queries from current `.sql` tests I wrote above.

This approach seems to be difficult to implement, because the **data cache method** should be determined at system startup (we need to create the corresponding FileCache Class from this value in the configuration file when the system is started). The cache algorithm should not be modified when the system is running. In short, the data cache method is a server-side global strategy, not determined by the client. If the data cache method needs to be modified, the user should stop the system and modify the data cache method in the configuration file, and then restart the system.

Reply:
Ah, right, then just. 2 storage policies in config. This will work.

Reply:
> I would also rather copy tests where cache is tested or add randomization, so both methods are tested in CI. Lets copy tests to include both policies. To make it easier, let's add a setting in Settings.h which will define a default cache policy. It will be used if there is no value `cache_method` in disk config (instead of hardcoded `"ARC"` in `config.getString(config_prefix + ".data_cache_method", "ARC");`). So let's add it for the following tests:
> 
> ```
> stateless:
> 02241_remote_filesystem_cache_on_insert.sql
> 02226_s3_with_cache.sql
> 02240_system_remote_filesystem_cache.sql
> 
> stateful:
> 00170_s3_cache.sql
> ```
> 
> Probably it could be helpful to make an `.sh` instead of these `.sql` tests (see `0_stateless`, there are many `.sh` tests). So in this `.sh` test there can be a loop over a list of cache policies (for now only two). For each policy it will run clickHouse-client command with `--queries-file`. In the file let's put the queries from current `.sql` tests I wrote above.

I think it is possible to achieve hot update of the cache method, but the cost is that it needs to lock and reload all cache resources according to the new cache method. During this period, all requests to access the cache will be blocked, and the current code will be changed greatly. Maybe we can implement this feature in a future version.

Reply:
> and the current code will be changed greatly

Not necessarily I think. At the moment disk config settings can be dynamically reloaded. So we can allow to change cache method too, so that for some period of time two caches can exist, but the new cache will be used for all new queries after settings are reloaded dynamically and the cache with previous method will just serve remaining queries and self-destruct when there are not more cells which are hold by anyone. 

But let's not implement this for now.

For tests two storage policies in config would suffice. In `.sh` file each of them can be chosen when a table is created. This will work for stateless tests only. For stateful will not work. Would be good to add randomization for stateful tests. I think not difficult to randomly choose storage policy in this script here https://github.com/ClickHouse/ClickHouse/blob/master/docker/test/stateful/run.sh#L98